File: skillcorpus_plugin/engine-typescript/src/marketplace-source.ts
/** Public ClawHub and skillhub.cn search/download adapters. */
import { existsSync } from 'node:fs'
import { readFile } from 'node:fs/promises'
import { join } from 'node:path'
import { bundleRoot, extractBundle } from './bundle.js'
import { compactCatalogQuery } from './relevance.js'
import type { RouterHit, SearchOptions, SkillSource } from './types.js'

export type MarketplaceKind = 'clawhub' | 'skillhub_cn'

export interface MarketplaceItem {
  id: string
  slug: string
  name: string
  description: string
  score: number
  owner?: string
  version?: string
  suspicious?: boolean
  installable?: boolean
  tags?: string[]
}

export class MarketplaceClient {
  readonly kind: MarketplaceKind
  private readonly base: string
  private readonly cacheDir: string
  private readonly timeoutMs: number
  private readonly downloadTimeoutMs: number

  constructor(kind: MarketplaceKind, endpoint: string, options: {
    cacheDir: string; timeoutMs?: number; downloadTimeoutMs?: number
  }) {
    this.kind = kind
    this.base = endpoint.replace(/\/+$/, '')
    this.cacheDir = options.cacheDir
    this.timeoutMs = options.timeoutMs ?? 5000
    this.downloadTimeoutMs = options.downloadTimeoutMs ?? 30_000
  }

  async search(query: string, signal?: AbortSignal, limit = 2): Promise<MarketplaceItem[]> {
    return this.kind === 'clawhub'
      ? this.searchClawHub(query, signal, limit)
      : this.searchSkillHubCn(query, signal, limit)
  }

  async install(hit: RouterHit, signal?: AbortSignal): Promise<{ dir: string; body: string }> {
    const slug = String(hit.meta.slug ?? hit.meta.id)
    const owner = String(hit.meta.owner ?? '')
    const version = String(hit.meta.version ?? 'v0')
    const key = `${this.kind}-${owner ? `${owner}_` : ''}${slug}@${version}`.replace(/[^A-Za-z0-9_.@-]+/g, '_')
    const destination = join(this.cacheDir, key)
    if (!existsSync(destination)) {
      const archive = await this.download(slug, owner, version, signal)
      await extractBundle(archive, destination)
    }
    const dir = await bundleRoot(destination)
    const skillMd = await readFile(join(dir, 'SKILL.md'), 'utf8')
    return { dir, body: stripFrontmatter(skillMd) }
  }

  private async searchClawHub(query: string, signal: AbortSignal | undefined, limit: number) {
    const url = new URL(`${this.base}/api/v1/search`)
    url.searchParams.set('q', query)
    url.searchParams.set('limit', String(limit))
    url.searchParams.set('nonSuspiciousOnly', 'true')
    const payload = await this.json(url, signal) as { results?: Record<string, unknown>[] }
    return (payload.results ?? []).flatMap(raw => {
      const slug = String(raw.slug ?? '')
      const native = raw.native as { skill?: Record<string, unknown> } | undefined
      const skill = native?.skill
      const trust = raw.trust as Record<string, unknown> | undefined
      if (!slug || trust?.visibility === 'blocked' || trust?.installability === 'blocked') return []
      return [{
        id: String(raw.id ?? slug), slug, name: String(raw.displayName ?? slug),
        description: String(raw.summary ?? skill?.summary ?? ''), score: Number(raw.score ?? 0),
        owner: String(raw.ownerHandle ?? ''), version: String(raw.version ?? skill?.latestVersionId ?? 'v0'),
        suspicious: skill?.isSuspicious === true, installable: trust?.installability == null || trust.installability === 'installable',
        tags: Array.isArray(skill?.topics) ? skill.topics.map(String) : [],
      }]
    })
  }

  private async searchSkillHubCn(query: string, signal: AbortSignal | undefined, limit: number) {
    const url = new URL(`${this.base}/api/skills`)
    url.searchParams.set('keyword', query)
    url.searchParams.set('sortBy', 'score')
    url.searchParams.set('order', 'desc')
    url.searchParams.set('page', '1')
    url.searchParams.set('pageSize', String(limit))
    const payload = await this.json(url, signal) as { code?: number; data?: { skills?: Record<string, unknown>[] } }
    if (payload.code !== 0) throw new Error('skillhub.cn search failed')
    return (payload.data?.skills ?? []).flatMap(raw => {
      const slug = String(raw.slug ?? '')
      if (!slug || malicious(raw.securityReports)) return []
      const namespace = raw.namespace as Record<string, unknown> | undefined
      return [{ id: String(namespace?.canonicalName ?? slug), slug,
        name: String(raw.name ?? slug), description: String(raw.description_zh ?? raw.description ?? ''),
        score: Number(raw.score ?? 0), owner: String(raw.ownerName ?? namespace?.handle ?? ''),
        version: String(raw.version ?? 'v0'), installable: true }]
    })
  }

  private async download(slug: string, owner: string, version: string, signal?: AbortSignal) {
    const url = new URL(`${this.base}/api/v1/download`)
    url.searchParams.set('slug', slug)
    if (this.kind === 'clawhub' && owner) url.searchParams.set('ownerHandle', owner)
    if (this.kind === 'skillhub_cn' && version !== 'v0') url.searchParams.set('version', version)
    url.searchParams.set('source', this.kind === 'skillhub_cn' ? 'dsh' : 'cli')
    return this.bytes(url, signal, this.downloadTimeoutMs)
  }

  private async json(url: URL, signal?: AbortSignal): Promise<unknown> {
    const bytes = await this.bytes(url, signal, this.timeoutMs)
    return JSON.parse(bytes.toString('utf8'))
  }

  private async bytes(url: URL, signal: AbortSignal | undefined, timeoutMs: number): Promise<Buffer> {
    const controller = new AbortController()
    const timer = setTimeout(() => controller.abort(), timeoutMs)
    const abort = () => controller.abort()
    if (signal?.aborted) controller.abort()
    signal?.addEventListener('abort', abort, { once: true })
    try {
      const response = await fetch(url, { signal: controller.signal })
      if (!response.ok) throw new Error(`${this.kind} returned HTTP ${response.status}`)
      return Buffer.from(await response.arrayBuffer())
    } finally {
      clearTimeout(timer)
      signal?.removeEventListener('abort', abort)
    }
  }
}

export class MarketplaceSkillSource implements SkillSource {
  readonly name: MarketplaceKind
  weight: number
  constructor(readonly client: MarketplaceClient, options: { weight?: number } = {}) {
    this.name = client.kind
    this.weight = options.weight ?? 0.75
  }
  async search(query: string, options: SearchOptions, k: number): Promise<RouterHit[]> {
    const items = await this.client.search(compactCatalogQuery(query), options.signal, Math.min(2, k))
    return items.filter(item => !item.suspicious && item.installable !== false).slice(0, Math.min(2, k)).map(item => ({
      qualifiedId: `${this.name}/${item.id}`, name: item.name, content: '', score: item.score,
      meta: { source: this.name, id: item.id, slug: item.slug, owner: item.owner,
        version: item.version, description: item.description, tags: item.tags },
    }))
  }
}

function malicious(value: unknown): boolean {
  if (!value || typeof value !== 'object') return false
  return Object.values(value).some(report => report && typeof report === 'object'
    && ['malicious', 'suspicious'].includes(String((report as Record<string, unknown>).status)))
}

function stripFrontmatter(text: string): string {
  if (!text.startsWith('---')) return text
  const end = text.indexOf('\n---', 3)
  return end < 0 ? text : text.slice(end + 4).replace(/^\n+/, '')
}

File: skillcorpus_plugin/engine-python/skillsearch/sources/marketplace_source.py
"""Public ClawHub and skillhub.cn adapters with safe bundle caching."""

from __future__ import annotations

import os
import re
import shutil
import uuid
from pathlib import Path
from typing import Any, Literal

import httpx

from skillsearch import provenance
from skillsearch.hub_client import SkillHubClient
from skillsearch.local_store import _parse_frontmatter
from skillsearch.provenance import identity
from skillsearch.types import RouterHit

MarketplaceKind = Literal["clawhub", "skillhub_cn"]


class MarketplaceClient:
    def __init__(
        self,
        kind: MarketplaceKind,
        endpoint: str,
        *,
        cache_dir: Path,
        # Set to the shared skills directory to install there instead of into
        # the cache. A skill under the cache is deliberately outside every
        # scan; one here is scanned, shared with the other hosts, and carries
        # a provenance marker so fusion knows it is the catalogue's own entry.
        install_root: Path | None = None,
        timeout_s: float = 5.0,
        download_timeout_s: float = 30.0,
        client: httpx.AsyncClient | None = None,
    ) -> None:
        self.kind = kind
        self._base = endpoint.rstrip("/")
        self._cache_dir = cache_dir
        self._install_root = install_root
        self._download_timeout_s = download_timeout_s
        self._owns_client = client is None
        self._client = client or httpx.AsyncClient(timeout=httpx.Timeout(timeout_s))

    async def aclose(self) -> None:
        if self._owns_client:
            await self._client.aclose()

    async def search(self, query: str, *, limit: int = 2) -> list[dict[str, Any]]:
        return await (
            self._search_clawhub(query, limit) if self.kind == "clawhub" else self._search_skillhub_cn(query, limit)
        )

    async def install(self, hit: RouterHit) -> dict[str, str]:
        slug = str(hit.meta.get("slug") or hit.meta.get("id"))
        owner = str(hit.meta.get("owner") or "")
        version = str(hit.meta.get("version") or "v0")
        if self._install_root is not None:
            return await self._install_shared(hit, slug, owner, version)
        key = re.sub(r"[^A-Za-z0-9_.@-]+", "_", f"{self.kind}-{owner + '_' if owner else ''}{slug}@{version}")
        destination = self._cache_dir / key
        was_cached = destination.exists()
        if not was_cached:
            staging = destination.with_name(f"{destination.name}.incoming-{os.getpid()}-{uuid.uuid4().hex[:8]}")
            try:
                SkillHubClient._safe_extract(await self._download(slug, owner, version), staging)
                try:
                    staging.rename(destination)
                except OSError:
                    if not destination.exists():
                        raise
                    shutil.rmtree(staging, ignore_errors=True)
            except BaseException:
                shutil.rmtree(staging, ignore_errors=True)
                raise
        try:
            root = SkillHubClient._bundle_root(destination)
            text = (root / "SKILL.md").read_text(encoding="utf-8")
            _, body = _parse_frontmatter(text)
            return {"dir": str(root), "skill_md": body}
        except (OSError, UnicodeError):
            # A directory is a cache hit only when it contains a readable
            # SKILL.md. Remove incomplete/corrupt entries so one interrupted
            # install cannot poison every later turn. Retry a pre-existing
            # cache once; a freshly downloaded bad archive still fails.
            shutil.rmtree(destination, ignore_errors=True)
            if was_cached:
                return await self.install(hit)
            raise

    async def _install_shared(self, hit: RouterHit, slug: str, owner: str, version: str) -> dict[str, str]:
        """Install into the shared directory, one directory per identity.

        Per identity rather than per version, unlike the cache: the shared
        directory is scanned, so ``x@1`` beside ``x@2`` would put two ranked
        copies of one skill in front of the model. An update replaces, and
        `swap_into_place` keeps the previous version live until the new one is
        complete.

        A same-version reinstall is skipped, so a retrieval that happens to hit
        an installed skill does not rewrite it every turn.
        """
        install_root = self._install_root
        if install_root is None:  # pragma: no cover - unreachable via `install`
            raise RuntimeError("no install root")
        marketplace_slug = f"{owner}_{slug}" if owner else slug
        dest = provenance.slug_dir(install_root, self.kind, marketplace_slug)
        existing = provenance.read_marker(dest)
        if existing is None or existing.version != version:
            staging = dest.with_name(f"{dest.name}.incoming-{os.getpid()}-{uuid.uuid4().hex[:8]}")
            try:
                SkillHubClient._safe_extract(await self._download(slug, owner, version), staging)
                body_root = SkillHubClient._bundle_root(staging)
                text = (body_root / "SKILL.md").read_text(encoding="utf-8")
                _, body = _parse_frontmatter(text)
                provenance.write_marker(
                    body_root,
                    provenance.Origin(
                        origin=provenance.identity(self.kind, marketplace_slug),
                        source=self.kind,
                        slug=marketplace_slug,
                        version=version,
                        sha256=provenance.body_digest(body),
                        installed_at=provenance.now(),
                    ),
                )
                provenance.swap_into_place(staging, dest)
            except BaseException:
                shutil.rmtree(staging, ignore_errors=True)
                raise
        root = SkillHubClient._bundle_root(dest)
        text = (root / "SKILL.md").read_text(encoding="utf-8")
        _, body = _parse_frontmatter(text)
        return {"dir": str(root), "skill_md": body}

    async def _search_clawhub(self, query: str, limit: int) -> list[dict[str, Any]]:
        response = await self._client.get(
            f"{self._base}/api/v1/search",
            params={
                "q": query,
                "limit": limit,
                "nonSuspiciousOnly": "true",
            },
        )
        response.raise_for_status()
        output = []
        for raw in (response.json() or {}).get("results", []):
            skill = (raw.get("native") or {}).get("skill") or {}
            trust = raw.get("trust") or {}
            slug = str(raw.get("slug") or "")
            if not slug or skill.get("isSuspicious") or trust.get("visibility") == "blocked":
                continue
            if trust.get("installability") not in (None, "installable"):
                continue
            output.append(
                {
                    "id": raw.get("id") or slug,
                    "slug": slug,
                    "name": raw.get("displayName") or slug,
                    "description": raw.get("summary") or skill.get("summary") or "",
                    "score": float(raw.get("score") or 0),
                    "owner": raw.get("ownerHandle") or "",
                    "version": raw.get("version") or skill.get("latestVersionId") or "v0",
                    "tags": skill.get("topics") or [],
                }
            )
        return output

    async def _search_skillhub_cn(self, query: str, limit: int) -> list[dict[str, Any]]:
        response = await self._client.get(
            f"{self._base}/api/skills",
            params={
                "keyword": query,
                "sortBy": "score",
                "order": "desc",
                "page": 1,
                "pageSize": limit,
            },
        )
        response.raise_for_status()
        payload = response.json() or {}
        if payload.get("code") != 0:
            raise RuntimeError("skillhub.cn search failed")
        output = []
        for raw in (payload.get("data") or {}).get("skills", []):
            if _malicious(raw.get("securityReports")):
                continue
            namespace = raw.get("namespace") or {}
            slug = str(raw.get("slug") or "")
            if not slug:
                continue
            output.append(
                {
                    "id": namespace.get("canonicalName") or slug,
                    "slug": slug,
                    "name": raw.get("name") or slug,
                    "description": raw.get("description_zh") or raw.get("description") or "",
                    "score": float(raw.get("score") or 0),
                    "owner": raw.get("ownerName") or namespace.get("handle") or "",
                    "version": raw.get("version") or "v0",
                    "tags": [],
                }
            )
        return output

    async def _download(self, slug: str, owner: str, version: str) -> bytes:
        params = {"slug": slug, "source": "dsh" if self.kind == "skillhub_cn" else "cli"}
        if self.kind == "clawhub" and owner:
            params["ownerHandle"] = owner
        if self.kind == "skillhub_cn" and version != "v0":
            params["version"] = version
        response = await self._client.get(
            f"{self._base}/api/v1/download",
            params=params,
            timeout=httpx.Timeout(self._download_timeout_s),
            follow_redirects=True,
        )
        response.raise_for_status()
        return response.content


class MarketplaceSkillSource:
    def __init__(self, client: MarketplaceClient, *, weight: float = 0.75) -> None:
        self.client = client
        self.name = client.kind
        self.weight = weight

    async def search(self, query: str, history: list[dict[str, Any]], k: int) -> list[RouterHit]:
        del history
        items = await self.client.search(query, limit=min(2, max(0, k)))
        return [
            RouterHit(
                qualified_id=f"{self.name}/{item['id']}",
                name=str(item["name"]),
                content="",
                score=float(item["score"]),
                # `origin` before the spread so a catalogue field of that
                # name cannot shadow the identity fusion collapses on.
                meta={"source": self.name, **item, "origin": identity(self.name, str(item.get("slug") or item["id"]))},
            )
            for item in items[: min(2, max(0, k))]
        ]


def _malicious(reports: Any) -> bool:
    return isinstance(reports, dict) and any(
        isinstance(report, dict) and report.get("status") in {"malicious", "suspicious"} for report in reports.values()
    )


__all__ = ["MarketplaceClient", "MarketplaceSkillSource"]
